Port dragnet to python 3

pakelley opened the pull request:
So this is a large PR with a lot of changes... Included in the PR is a completion of @bdewilde's changes in #41, as well as porting the library to Python 3. Also, as it seems like the library is moving toward the changes made in the other open PR, we branched off of that instead of master. Here's a somewhat in-depth overview:

# Changes
* I added some type-conversion functions in `compat.py`
  * docstrings included, so hopefully they're simple to use/understand :)
  * it would be nice to use something like `future` here, but it was giving me more hassle than just using my own functions
  * In particular, I ran into problems w/ `str(bytes('a string')) == "b'a string'"`
* conda environments
  * env file added, and incorporated into `.travis.yml`
* `blocks.pyx`
  * My basic policy here was to convert to bytes on the way into any Cython files, and to a string on the way out
  * Aside from that idea, it's all mostly mechanical; just convert when appropriate
* `test_kohlschuetter.py` has been adjusted to ensure explicit unicode chars are properly interpreted
* Model tests have been re-introduced (they were skipped over in the other PR)
* Pre-trained models have been added for sklearn>=0.18

# Slightly Questionable Things
* I simplified the chardet call in `blocks.pyx`
  * From my understanding, the pkg-level call is just sugar for what was there before
* I added cythonize to `setup.py`
  * Gives access to some niceties(e.g. GDB)
* It seems that the new API for Extractor doesn't support returning both comments and content as separate lists
  * The model tests now make an explicit call to each type of extraction rather than doing it all in one go
  * If returning both is implemented, I'm more than happy to update this :)
* I increased `rtol` in `np.allclose` in `test_readability.py`
  * It's still a reasonable tolerance, and the test is referencing a saved/hard-coded file for the comparison values
* f1 threshold decreased in `test_model.py`
  * I haven't had the chance to train a model that fits the expected accuracy
  
# Questions
* I included pre-trained models for sklearn>=0.18, but would it maybe make more sense to just include the params to train those models? And maybe at most a simple script to execute them?
* These are mentioned above, but I would love to have some feedback on the changes to `rtol` in the readability tests, and on the f1 threshold in the model tests.
  * I'm also curious about how you trained parameters, as I've done a bit of gridsearching and haven't found params that get very close to a .98 f1 score (I think the best one I've gotten was around .91, but I haven't been able to do any better than that)

For what it's worth, this is my first PR to an open-source project, so be gentle :)

matt-peters commented:
Great to see this!!  Thank you!  I'm traveling for a few weeks so won't be able to look at this in the near term, but will definitely take a look when I can, or failing that when I return.  In the meantime, someone else from Moz might review.

bdewilde commented:
Hey folks, just wondering if there's been any progress on reviewing this PR. (Looks like no...?) Is there anything I can do to assist? Am willing to review, if you want my eyes on this.

matt-peters commented:
This PR is massive :-) and it's great to see the Python 3.5 tests passing in Travis.

@bdewilde I'd love another set of eyes on this, especially as also includes the completion of your PR.

I'll try to look through this early next week -- but overall I'm less concerned about the API changes (since they include those in the former PR and those seemed sensible) then the overall performance.

Has anyone benchmarked this on the dragnet-data test set (both model F1 and speed)?  I don't expect there to be a big performance drop moving to Python 3 but if there is then it's good to know about and try to track down.  `blocks.pyx` does a lot of low level string handling, some pieces of which were performance sensitive.

donguklim commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+    if PY2:
+        model_path = 'py2_sklearn_0.15.2_0.17.1'
+    else:
+        model_path = 'py3_sklearn_0.15.2_0.17.1'
```
This model path does not exist. Causing error if sklearn version is from 0.152 to 0.17.1 while using python3

pakelley replied:
Sorry about that, I missed this file :) Added/pushed it

donguklim commented on dragnet/util.py:
```diff
@@ -90,20 +97,75 @@ def evaluation_metrics(predicted, actual, bow=True):
+            pkg_dirname = pkgutil.get_loader('dragnet').name
+        dirname = os.path.join(pkg_dirname, 'pickled_models', model_path)
+    filepath = os.path.join(dirname, filename)
+    return joblib.load(filepath)
```
f
You should use pkgutil.get_data function to get the data file directory.
This only gives relative directory and causes file does not exist exception if I try to use the library after installing with the make file.

I changed line 164 to 170 as shown in bellow in my local repository. 
    if dirname is None:
        pkg_name = pkgutil.get_loader('dragnet').name
        pkg_data_path = os.path.join('pickled_models', model_path, filename)
        filepath = pkgutil.get_data(pkg_name, pkg_data_path)
    else:
        filepath = os.path.join(dirname, filename)

donguklim commented on dragnet/util.py:
```diff
@@ -90,20 +97,75 @@ def evaluation_metrics(predicted, actual, bow=True):
+            pkg_dirname = pkgutil.get_loader('dragnet').filename
+        else:
+            pkg_dirname = pkgutil.get_loader('dragnet').name
+        dirname = os.path.join(pkg_dirname, 'pickled_models', model_path)
```
This only gives relative path within dragnet package, which causes file not exist exception.
I modified line 165 to 168 as shown bellow

pkg_filename = pkgutil.get_loader('dragnet').get_filename()
pkg_dirname = os.path.dirname(pkg_filename)
dirname = os.path.join(pkg_dirname, 'pickled_models', model_path)

I have not tested this on Python2, but when I checked python 2 document, python2 also have the same functions os.path.dirname, and pkgutil.get_loader().get_filename()

pakelley replied:
yep, that works. I tested this on Python 2 as well, and updated this code accordingly

nehalecky commented:
@pakelley, could you comment on @donguklim reviews above? 

@bdewilde, were you ever able to review? We'd really appreciate it, as we'd really like to get feedback and a merge for new shiny py3-supported release!

pakelley commented:
@nehalecky sure thing, made some updates wrt @donguklim 's suggestions

I'll keep an eye out for more updates here

nehalecky commented:
@pakelley thanks much!

@matt-peters, regarding https://github.com/seomoz/dragnet/pull/52#issuecomment-341789256
> Has anyone benchmarked this on the dragnet-data test set (both model F1 and speed)? I don't expect there to be a big performance drop moving to Python 3 but if there is then it's good to know about and try to track down. blocks.pyx does a lot of low level string handling, some pieces of which were performance sensitive.

We're happy to do this, but would like to have the baseline benchmarks and environment configuration to work from. Would you or someone else be able to help us with this?

Thanks again!

matt-peters commented:
Great to see everyone working together on this PR again!

@nehalecky -- [this blog post](https://moz.com/devblog/benchmarking-python-content-extraction-algorithms-dragnet-readability-goose-and-eatiht/) has some benchmarks for the Python 2 version.  The two important numbers are the overall F1 on the dragnet training data (https://github.com/seomoz/dragnet_data) and the throughput (documents / second).  Those numbers were from a 2014 era Macbook so any modern-ish CPU would be fine.  We don't need to reproduce them exactly, just as long as there isn't a significant performance regression.

nehalecky commented:
@matt-peters, Happy New Year! Thanks for the helpful reply, we're going to begin benchmarking the port today. One question, are you still a maintainer of this code base, and able to approve and merge the PR? We're eager to see momentum behind this!

bdewilde commented:
Hi all, a thousand apologies that I've dropped the ball and haven't yet reviewed the later changes — maybe this can be my first completed new years' resolution? 😄  Is there anything in particular that you want my eyes on, or just give the whole thing a once-over?

nehalecky commented:
hey @bdewilde! I think it's a great resolution, and would love if you could review: 
>  just give the whole thing a once-over
preferably this first!

matt-peters commented:
@nehalecky 
I am still the maintainer and can merge the PR and push a new version to PyPI.  I'm on vacation this week so may be a little delayed (especially to push an update to PyPI) but will be back in the office next week.

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+
+    Parameters
+    ----------
+    maybe_bytes : if this is a bytes-like object, it will be converted to a utf-8 string
```
since `encoding` is a parameter, this doesn't necessarily have to be converted to a utf-8 string, right? if so, the wording should probably reflect it.

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+
+    Parameters
+    ----------
+    maybe_str : if this is a string-like object, it will be converted to bytes(assumes utf-8 encoding)
```
same comment as above re: utf-8 not required

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+    Parameters
+    ----------
+    list_ : list
+        any bytes-like objects contained in the list will be converted to strings (assumed utf-8 encoding)
```
_cough_ looks like all of the cast functions have the same comment about utf-8

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+        encoding: str, default: 'utf-8'
+            encoding to be used when decoding bytes
+    """
+    old_keys = dict_.keys()
```
Nit: I don't think there's a need to assign these two old_ variables, since you can just do, e.g. `new_keys = str_list_cast(dict_.keys(), **kwargs) if include_keys else dict_.keys()`. Am assuming there's no real performance difference, I just found it more confusing to read the code.

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+    old_vals = dict_.values()
+    new_keys = str_list_cast(old_keys, **kwargs) if include_keys else old_keys
+    new_vals = str_list_cast(old_vals, **kwargs) if include_vals else old_vals
+    new_dict = dict(zip(new_keys, new_vals))
```
This should probably be `zip_`, no?

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+    return new_dict
+
+
+def bytes_dict_cast(dict_, include_keys=True, include_vals=True, **kwargs):
```
same comments as in the function above

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+        encoding: str, default: 'utf-8'
+            encoding to be used when decoding bytes
+    """
+    block.text = str_cast(block.text, **kwargs)                    if include_text        else block.text
```
Using whitespace like this is strongly discouraged in Python. Also, in the case that `include_*` is False, there's no need to assign the value back to itself, right? So maybe this would be both more readable and more performant:

```
if include_text:
    block.text = str_cast(block.text, **kwargs)
if include_link_tokens:
    block.link_tokens = str_list_cast(block.link_tokens, **kwargs)
...
```

bdewilde commented on dragnet/compat.py:
```diff
@@ -8,20 +8,259 @@
+    return block
+
+
+def bytes_block_cast(block,
```
same comments as in the function above

bdewilde commented on dragnet/util.py:
```diff
@@ -38,7 +45,7 @@ def dameraulevenshtein(seq1, seq2):
     # so we only store those.
     oneago = None
-    thisrow = range(1, len(seq2) + 1) + [0]
+    thisrow = list(range(1, len(seq2) + 1)) + [0]
```
`range_`?

pakelley commented:
I've updated this PR off of @bdewilde 's comments, as well as some bottlenecks from the `compat` functions I added earlier(this was mainly just making less calls in the middle of the extraction process).

For a model with an F1-score of 0.86 extraction still runs a bit slower than master, at 53 docs/sec on Python 3 and 31 docs/sec on Python 2. This is an improvement from the previous 35 docs/sec on Python 3 and 20 docs/sec on Python 2. On master, extraction runs for me at about 70 docs/sec.

With regard to the `chardet` calls, I compared the throughput when instantiating a single UniversalDetecotr(as mentioned by @bdewilde ) and there didn't seem to be a difference, so I just left the `detect` call as it was.

Hopefully this PR looks a bit better now :) Let me know what you all think of these benchmarks/changes.

bdewilde commented:
Just to be clear: These numbers are to be compared to (F1, docs/sec) pairs of (0.838, 70) for content only and (0.848, 75) for content and comments? I've pulled them from that 2015 blog post. What's the F1 score you get using the default setting on master? I'd feel better about the slightly slower classification knowing, for sure, that at least it's more accurate.

The slowness isn't a _huge_ deal, especially since it's still faster than readability and goose, but may be something to profile and improve in a subsequent PR... :)

pakelley commented:
Oh yeah, great question! This is all using content and comments. Using the default model on master I get an F1 of 0.85, but using my model I get 0.82, so the code in this PR is slightly more accurate for my model at least. I'm not sure what the params are for the model loaded by default on master, and I haven't been able to reproduce it either.

Also, fwiw, it occurred to me that I was using a different blockifier than I was on master. After switching that I get an F1-score of 0.85 on Python 3 at 60 docs/sec. On Python 2 I get the same F1, but still at 31-32 docs/sec.

nehalecky commented on .travis.yml:
```diff
@@ -1,11 +1,54 @@
+before_install:
+  - ls /etc/profile.d/
+
+install:
```
hi @bdewilde !
We use conda internally for managing our packaging and dependency management, and find it way more sane than where historical python buildtools. FWIW, [conda](https://conda.io/docs/) is developed by [Anaconda Inc](https://www.anaconda.com/) (formerly Continuum Analytics), and it's used heavily across scientific communities. 

You can read about some differences it has vs pips here:
http://jakevdp.github.io/blog/2016/08/25/conda-myths-and-misconceptions/

and definitely you should try it out—HTH!

bdewilde commented:
Hey folks, what's the status on this PR now? Are we good to go? If not, what's left?

pakelley commented:
Good to go on my end, fwiw :) Also, I ran the same benchmarks on content-only extraction the other day and got similar results. I can re-run and post the numbers if that's a concern.

dlecocq commented:
I don't know who on the Moz side is managing this repo since both @matt-peters and I have long since left. Maybe @b4hand could take a look?

b4hand commented:
I'm not really familiar at all with this code. I'd prefer to wait for confirmation from @matt-peters, but if he doesn't respond, I can merge this on Friday since it seems that the general consensus is that this works.

nehalecky commented:
@b4hand, sounds great. Going forward, perhaps Moz / project owners can consider adding a few additional project maintainers? [We'd be](https://github.com/bomboradata) happy happy to help out. :)

b4hand commented:
Yeah, unfortunately, this project could probably be spun off into its own org. I don't think we have any active maintainers for it internally anymore.

nehalecky commented:
@b4hand, that feels right, and we'd be happy to support. Is the change to org that something you could do?

b4hand commented:
I have the ability to make the org change, but I'll need to get approval.

nehalecky commented:
@b4hand, sounds great! We look forward to supporting dragnet into the future with Python 3 and wherever else we might go!

matt-peters commented on dragnet/models.py:
```diff
@@ -1,54 +0,0 @@
-        compressed='gzip')
-
-    # monkey patches
-    content_extractor._blockifier = TagCountNoCSSReadabilityBlockifier
```
This is where the `_blockifier` was monkey patched when loading models.

pakelley replied:
I don't know too much about this change, as it was made in @bdewilde's branch, but I would be happy to train a new model that simply uses the `TagCountNoCSSReadabilityBlockifier` if that helps. FWIW, on Python 3 and sklearn 0.19, the model loads with `TagCountReadabilityBlockifier`.

matt-peters replied:
I think that's worthwhile to try.  If I remember correctly, the CSS features are pretty slow and don't add much to the overall performance so dropping them and using the `TagCountNoCSSReadabilityBlockifier` usually gives the best accuracy/speed tradeoff.

pakelley replied:
Sounds good, I'll get some new models up in a new PR as soon as I have the chance

matt-peters commented on dragnet/util.py:
```diff
@@ -90,20 +97,73 @@ def evaluation_metrics(predicted, actual, bow=True):
+        pkg_dirname = os.path.dirname(pkg_filename)
+        dirname = os.path.join(pkg_dirname, 'pickled_models', model_path)
+    filepath = os.path.join(dirname, filename)
+    return joblib.load(filepath)
```
Is it necessary to also patch in the blockifier here when loading?

matt-peters commented:
As far as org changes: I'd support an organization change if Moz will also support it.  I don't know whether Moz would want to maintain some level of admin/write access to the repo?  Since there are several different organizations, transferring ownership to a 3rd party github account then adding interested parties with write/admin access might make sense.
